File: backend/services/openai_service.py
"""OpenAI service wrapper for recipe JSON generation."""

import json
import os

from openai import OpenAI


class OpenAIConfigError(Exception):
    """Raised when OpenAI configuration is missing."""


class OpenAIJSONError(Exception):
    """Raised when OpenAI response is not valid JSON."""


class OpenAIService:
    def __init__(self) -> None:
        self.api_key = os.getenv("OPENAI_API_KEY", "")
        self.model = os.getenv("OPENAI_MODEL", "gpt-4o-mini")

    def _get_client(self) -> OpenAI:
        if not self.api_key:
            raise OpenAIConfigError(
                "OpenAI is not configured. Missing OPENAI_API_KEY in backend/.env."
            )
        return OpenAI(api_key=self.api_key)

    def generate_recipe_json(self, prompt: str) -> dict:
        client = self._get_client()
        try:
            response = client.chat.completions.create(
                model=self.model,
                messages=[
                    {"role": "system", "content": "Return strict JSON only."},
                    {"role": "user", "content": prompt},
                ],
                response_format={"type": "json_object"},
            )
            content = response.choices[0].message.content
            if not content:
                raise OpenAIJSONError("OpenAI returned empty content.")
            return json.loads(content)
        except json.JSONDecodeError as exc:
            raise OpenAIJSONError(f"OpenAI returned invalid JSON: {exc}")
        except OpenAIConfigError:
            raise
        except OpenAIJSONError:
            raise
        except Exception as exc:
            raise RuntimeError(f"OpenAI request failed: {type(exc).__name__}: {exc}")

File: supabase/functions/api/services/openaiService.ts
export class OpenAIConfigError extends Error {}
export class OpenAIRequestError extends Error {}
export class OpenAIJSONError extends Error {}

export function getOpenAIModel(): string {
  return (Deno.env.get("OPENAI_MODEL") || "gpt-4o-mini").trim() || "gpt-4o-mini";
}

export async function generateRecipeWithOpenAI(systemPrompt: string, userPrompt: string): Promise<Record<string, unknown>> {
  const apiKey = Deno.env.get("OPENAI_API_KEY")?.trim();
  if (!apiKey) throw new OpenAIConfigError("Missing OPENAI_API_KEY for Edge Function OpenAI generation.");

  const model = getOpenAIModel();
  const res = await fetch("https://api.openai.com/v1/chat/completions", {
    method: "POST",
    headers: {
      "Content-Type": "application/json",
      Authorization: `Bearer ${apiKey}`,
    },
    body: JSON.stringify({
      model,
      response_format: { type: "json_object" },
      temperature: 0.4,
      messages: [
        { role: "system", content: systemPrompt },
        { role: "user", content: userPrompt },
      ],
    }),
  });

  if (!res.ok) throw new OpenAIRequestError(`OpenAI request failed with status ${res.status}.`);

  const data = await res.json().catch(() => null) as Record<string, unknown> | null;
  const choices = (data?.choices as Array<Record<string, unknown>> | undefined) || [];
  const content = choices[0]?.message && (choices[0].message as Record<string, unknown>).content;

  if (!content || typeof content !== "string") throw new OpenAIRequestError("OpenAI returned empty response.");

  try {
    return JSON.parse(content) as Record<string, unknown>;
  } catch {
    throw new OpenAIJSONError("OpenAI returned invalid JSON.");
  }
}
